mephisto/abstractions/databases/local_singleton_database.py

```diff
@@ -128,3 +129,93 @@ def new_agent(
         unit.db_status = AssignmentState.ASSIGNED
         unit.worker_id = agent.worker_id
         return agent_id
+
+    def find_units(
+        self,
+        task_id: Optional[str] = None,
+        task_run_id: Optional[str] = None,
+        requester_id: Optional[str] = None,
+        assignment_id: Optional[str] = None,
+        unit_index: Optional[int] = None,
+        provider_type: Optional[str] = None,
+        task_type: Optional[str] = None,
+        agent_id: Optional[str] = None,
+        worker_id: Optional[str] = None,
+        sandbox: Optional[bool] = None,
+        status: Optional[str] = None,
+    ) -> List[Unit]:
+        """
+        Uses caching to offset the cost of the most common queries. Defers
+        to the underlying DB for anything outside of those cases.
+        """
+
+        # Finding units is the most common small DB call to be optimized, as
+        # every assignment has multiple units. Thus, we try to break up the
+        # units to be queried by assignment, ensuring the most commonly
+        # queried edge is improved.
+        if assignment_id is not None:
+            if all(
+                v is None
+                for v in [
+                    task_id,
+                    task_run_id,
+                    requester_id,
+                    unit_index,
+                    provider_type,
+                    task_type,
+                    agent_id,
+                    worker_id,
+                    sandbox,
+                    status,
+                ]
+            ):
+                units = self._assignment_to_unit_mapping.get(assignment_id)
+                if units is None:
+                    units = super().find_units(assignment_id=assignment_id)
+                    self._assignment_to_unit_mapping[assignment_id] = units
+                return units
+
+        # Any other cases are less common and more complicated, and so we don't cache
+        return super().find_units(
+            task_id=task_id,
+            task_run_id=task_run_id,
+            requester_id=requester_id,
+            assignment_id=assignment_id,
+            unit_index=unit_index,
+            provider_type=provider_type,
+            task_type=task_type,
+            agent_id=agent_id,
+            worker_id=worker_id,
+            sandbox=sandbox,
+            status=status,
+        )
+
+    def new_unit(
+        self,
+        task_id: str,
+        task_run_id: str,
+        requester_id: str,
+        assignment_id: str,
+        unit_index: int,
+        pay_amount: float,
+        provider_type: str,
+        task_type: str,
+        sandbox: bool = True,
+    ) -> str:
+        """
+        Create a new unit with the given index. Raises EntryAlreadyExistsException
+        if there is already a unit for the given assignment with the given index.
+        """
+        if assignment_id in self._assignment_to_unit_mapping:
+            del self._assignment_to_unit_mapping[assignment_id]
```

Comment:
nit: I usually prefer popping dict keys instead of del. I am a bit scared of introducing bugs that may remove the whole dict instead of that single entry.

Reply:
I tend to prefer `del` as explicitly not caring about the value, `pop` if the value at the entry is important.